tests/kernel/interrupt/src/interrupt_offload.c

```diff
@@ -0,0 +1,266 @@
+{
+	printk("doing offload job\n");
+
+	if (offload_job_prio_higher) {
+		/*TESTPOINT: offload thread run right after irq end.*/
+		zassert_equal(orig_t_keep_run, 0,
+			"the offload did not run immediately.");
+	} else {
+		/*TESTPOINT: original thread run right after irq end.*/
+		zassert_equal(orig_t_keep_run, 1,
+			"the offload did not run immediately.");
+	}
+
+	k_sem_give(&end_sem);
+}
+
+/* offload work to work queue */
+void isr_handler(const void *param)
+{
+	struct k_work *work = ((struct interrupt_param *)param)->work;
+
+	zassert_not_null(work, "NULL");
+
+	k_work_init(work, entry_offload_job);
+
+	orig_t_keep_run = 0;
+
+	k_work_submit_to_queue(&wq_queue, work);
+}
+
+#if defined(CONFIG_DYNAMIC_INTERRUPTS)
+/*
+ * So far, we only test x86 and arch posix by real dynamic interrupt.
+ * Other arch will be add later.
+ */
+#if defined(CONFIG_X86)
+#define IV_IRQS 32
+#define TEST_IRQ_DYN_LINE 17
+#define TRIGGER_IRQ_DYN_LINE (TEST_IRQ_DYN_LINE + IV_IRQS)
+
+#elif defined(CONFIG_ARCH_POSIX)
+#define TEST_IRQ_DYN_LINE 5
+#define TRIGGER_IRQ_DYN_LINE 5
+
+#else
+#define TEST_IRQ_DYN_LINE -1
+#define TRIGGER_IRQ_DYN_LINE -1
+#endif
+
+#endif
+
+static void init_dyn_interrupt(void)
+{
+	/* If we cannot get a dynamic interrupt, skip test. */
+	if (TEST_IRQ_DYN_LINE == -1) {
+		ztest_test_skip();
+	}
+
+	int vector_num = irq_connect_dynamic(TEST_IRQ_DYN_LINE, 1,
+					isr_handler, (void *)&irq_param, 0);
+
+	printk("irq(%d)\n", vector_num);
+	zassert_true(vector_num > 0, "no vector can be used");
+	irq_enable(TEST_IRQ_DYN_LINE);
+}
+
+static void trigger_offload_interrupt(const bool real_irq, void *work)
+{
+	irq_param.work = work;
+
+	if (real_irq) {
+		trigger_irq(TRIGGER_IRQ_DYN_LINE);
+	} else {
+		irq_offload((irq_offload_routine_t)&isr_handler, &irq_param);
+	}
+}
+
+static void t_running(void *p1, void *p2, void *p3)
+{
+	k_sem_give(&start_sem);
+
+	while (1) {
+		orig_t_keep_run = 1;
+		k_usleep(1);
+	}
+}
+
+static void run_test_offload(int case_type, int real_irq)
+{
+	static bool wq_already_start;
+	int thread_prio = 1;
+
+	printk("case %d\n", case_type);
+
+	/* two semaphore used to sync start and end */
+	k_sem_init(&start_sem, 0, 1);
+	k_sem_init(&end_sem, 0, NUM_WORK);
+
+	if (offload_job_prio_higher) {
+		thread_prio = 2;
+	}
+
+	if (real_irq) {
+		init_dyn_interrupt();
+	}
+
+	k_tid_t tid = k_thread_create(&tdata, tstack, STACK_SIZE,
+			(k_thread_entry_t)t_running,
+			NULL, NULL, NULL,
+			K_PRIO_PREEMPT(thread_prio),
+			K_INHERIT_PERMS, K_NO_WAIT);
+
+	/* start a work queue thread if not existing */
+	if (!wq_already_start) {
+		k_work_queue_start(&wq_queue, wq_stack, STACK_SIZE,
+		1, NULL);
+
+		wq_already_start = true;
+	}
+
+	/* wait for thread start */
+	k_sem_take(&start_sem, K_FOREVER);
+
+	for (int i = 0; i < NUM_WORK; i++) {
+
+		switch (case_type) {
+		case TEST_OFFLOAD_MULTI_JOBS:
+			trigger_offload_interrupt(real_irq,
+					(void *)&offload_work[i]);
+		break;
+		case TEST_OFFLOAD_INDETICAL_JOBS:
```

Comment:
Here is the same misprint, IDENTICAL.

Reply:
Thanks for correcting this. Got it fixed.